File: binarysearch.io/353_tree_from_inorder_preorder.py
# class Tree:
#     def __init__(self, val, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def solve(self, preorder, inorder):
        pi = 0 
        n = len(preorder)
        def helper(i,j):
            nonlocal pi 
            if i > j:
                return 
            node = Tree(preorder[pi])
            pi += 1 
            for k in range(i,j+1):
                if inorder[k]==node.val:
                    node.left = helper(i,k-1)
                    node.right = helper(k+1,j)
                    break 
            return node 
        return helper(0,n-1)

File: binarysearch.io/353_tree_from_inorder_preorder.cpp
/**
 * class Tree {
 *     public:
 *         int val;
 *         Tree *left;
 *         Tree *right;
 * };
 */
Tree* helper(int &pi, int i, int j, vector<int>& preorder, vector<int>& inorder){
    if(i > j) return NULL;
    Tree* node = new Tree();
    node->val = preorder[pi];
    pi+= 1;
    for(int k = i ; k <=j; k++){
        if(inorder[k]==node->val){
            node->left = helper(pi,i,k-1,preorder,inorder);
            node->right = helper(pi,k+1,j,preorder,inorder);
            break;
        }
    }
    return node;
}
Tree* solve(vector<int>& preorder, vector<int>& inorder) {
    int pi = 0;
    return helper(pi,0,preorder.size()-1,preorder,inorder);
}